File: 11-container-with-most-water/container-with-most-water.java
class Solution {
    public int maxArea(int[] height) {
        int n = height.length;
        // right - left + 1 is width
        // min of right and left is height
        // where to move pointer?
        // got an idea of problem in 2 minutes after reading it

        int left = 0;
        int right = n-1;
        // as values are positive
        int maxArea = -10;
        while(left < right){
            int h = Integer.min(height[left],height[right]);
            int w = right - left;

            maxArea = Integer.max(h*w, maxArea);
            // want to maximize the area

            if( height[left] < height[right]){
                left+=1; // greedy for big value at right
            }
            else {
                right-=1; // greedy for big value at left
            }


        }

        return maxArea;
        
    }
}

File: 11-container-with-most-water/container-with-most-water.ts
function maxArea(height: number[]): number {
    let left = 0;
    let right = height.length - 1;
    // width right - left
    // height is current element
    let area= 0;
    let maxArea=0;
    while ( left < right ) {
        let heightDiff = Math.min( height[left], height[right]);
        let width = right - left;
        area = heightDiff  * width;
        maxArea = Math.max( area, maxArea);
        // trap by minimum
        // 8,7 --> 7
        if ( height[right] > height[left] ){
            left+=1;
        }
        else {
            right-=1;
        }
    }

    return maxArea;
    
};